### ISSUE TITLE
Unable to install term2 on Atom version 0.178.0

### ISSUE THREAD

**Issue by mdixon47** [webBoxio/atom-term2#90]
Does anyone have an idea on how to repair this problem? I have grown attached to the plugin and I would appreciate if someone can assist me in getting this fixed. 

> pty.js@0.2.5 install /private/var/folders/2r/twxtyrm12ks7x98wpyccvdk00000gn/T/apm-install-dir-115111-4598-p25mkv/node_modules/term2/node_modules/pty.js
> node-gyp rebuild

  CXX(target) Release/obj.target/pty/src/unix/pty.o

In file included from ../src/unix/pty.cc:16:
../node_modules/nan/nan.h:67:21: error: no type named 'ExternalAsciiStringResource' in 'v8::String'; did you mean 'ExternalStringResource'?
typedef v8::String::ExternalAsciiStringResource
        ~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~
                    ExternalStringResource
/Users/malikdixon/.atom/.node-gyp/.node-gyp/0.21.0/deps/v8/include/v8.h:1936:19: note: 'ExternalStringResource' declared here
  class V8_EXPORT ExternalStringResource
                  ^
In file included from ../src/unix/pty.cc:16:
In file included from ../node_modules/nan/nan.h:74:
../node_modules/nan/nan_new.h:164:26: error: class member cannot be redeclared
  static inline return_t New(NanExternalOneByteStringResource * value);
                         ^
../node_modules/nan/nan_new.h:163:26: note: previous declaration is here
  static inline return_t New(v8::String::ExternalStringResource * value);
                         ^
In file included from ../src/unix/pty.cc:16:
In file included from ../node_modules/nan/nan.h:74:
In file included from ../node_modules/nan/nan_new.h:190:
../node_modules/nan/nan_implementation_12_inl.h:222:22: error: redefinition of 'New'
Factory<v8::String>::New(NanExternalOneByteStringResource * value) {
                     ^
../node_modules/nan/nan_implementation_12_inl.h:217:22: note: previous definition is here
Factory<v8::String>::New(v8::String::ExternalStringResource * value) {
                     ^
In file included from ../src/unix/pty.cc:16:
In file included from ../node_modules/nan/nan.h:74:
../node_modules/nan/nan_new.h:319:1: error: redefinition of 'NanNew'
NanNew(NanExternalOneByteStringResource * value) {
^
../node_modules/nan/nan_new.h:313:1: note: previous definition is here
NanNew(v8::String::ExternalStringResource * value) {
^
In file included from ../src/unix/pty.cc:16:
../node_modules/nan/nan.h:1790:17: error: no member named 'GetExternalAsciiStringResource' in 'v8::String'; did you mean 'GetExternalOneByteStringResource'?
    return str->GetExternalAsciiStringResource();
                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                GetExternalOneByteStringResource
/Users/malikdixon/.atom/.node-gyp/.node-gyp/0.21.0/deps/v8/include/v8.h:2003:40: note: 'GetExternalOneByteStringResource' declared here
  const ExternalOneByteStringResource* GetExternalOneByteStringResource() const;
                                       ^
In file included from ../src/unix/pty.cc:16:
../node_modules/nan/nan.h:1790:12: error: cannot initialize return object of type 'const NanExternalOneByteStringResource *' (aka 'const v8::String::ExternalStringResource *') with an rvalue of type 'const v8::String::ExternalOneByteStringResource *'
    return str->GetExternalAsciiStringResource();
           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../node_modules/nan/nan.h:1800:17: error: no member named 'IsExternalAscii' in 'v8::String'; did you mean 'IsExternal'?
    return str->IsExternalAscii();
                ^~~~~~~~~~~~~~~
                IsExternal
/Users/malikdixon/.atom/.node-gyp/.node-gyp/0.21.0/deps/v8/include/v8.h:1900:8: note: 'IsExternal' declared here
  bool IsExternal() const;
       ^
In file included from ../src/unix/pty.cc:16:
../node_modules/nan/nan.h:1825:11: error: assigning to 'const char *' from incompatible type 'const uint16_t *' (aka 'const unsigned short *')
    *data = ext->data();
          ^ ~~~~~~~~~~~
../node_modules/nan/nan.h:1868:10: error: no matching function for call to 'Encode'
  return node::Encode(
         ^~~~~~~~~~~~
/Users/malikdixon/.atom/.node-gyp/.node-gyp/0.21.0/src/node.h:272:34: note: candidate function not viable: cannot convert argument of incomplete type 'const void *' to 'const char *'
NODE_EXTERN v8::Local<v8::Value> Encode(v8::Isolate* isolate,
                                 ^
/Users/malikdixon/.atom/.node-gyp/.node-gyp/0.21.0/src/node.h:278:34: note: candidate function not viable: requires 3 arguments, but 4 were provided
NODE_EXTERN v8::Local<v8::Value> Encode(v8::Isolate* isolate,
                                 ^
/Users/malikdixon/.atom/.node-gyp/.node-gyp/0.21.0/src/node.h:283:45: note: candidate function not viable: requires at most 3 arguments, but 4 were provided
                inline v8::Local<v8::Value> Encode(
                                            ^
/Users/malikdixon/.atom/.node-gyp/.node-gyp/0.21.0/src/node.h:64:53: note: expanded from macro 'NODE_DEPRECATED'
#define NODE_DEPRECATED(msg, fn) V8_DEPRECATED(msg, fn)
                                                    ^
/Users/malikdixon/.atom/.node-gyp/.node-gyp/0.21.0/deps/v8/include/v8config.h:329:45: note: expanded from macro 'V8_DEPRECATED'
# define V8_DEPRECATED(message, declarator) declarator
                                            ^
9 errors generated.
make: *** [Release/obj.target/pty/src/unix/pty.o] Error 1
gyp ERR! build error
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/Applications/Atom.app/Contents/Resources/app/apm/node_modules/npm/node_modules/node-gyp/lib/build.js:267:23)
gyp ERR! stack     at ChildProcess.emit (events.js:98:17)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (child_process.js:820:12)
gyp ERR! System Darwin 14.1.0
gyp ERR! command "node" "/Applications/Atom.app/Contents/Resources/app/apm/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /private/var/folders/2r/twxtyrm12ks7x98wpyccvdk00000gn/T/apm-install-dir-115111-4598-p25mkv/node_modules/term2/node_modules/pty.js
gyp ERR! node -v v0.10.35
gyp ERR! node-gyp -v v1.0.2
gyp ERR! not ok
npm ERR! Darwin 14.1.0
npm ERR! argv "/Applications/Atom.app/Contents/Resources/app/apm/bin/node" "/Applications/Atom.app/Contents/Resources/app/apm/node_modules/npm/bin/npm-cli.js" "--globalconfig" "/Applications/Atom.app/Contents/Resources/app/apm/.apmrc" "--userconfig" "/Users/malikdixon/.atom/.apmrc" "install" "/private/var/folders/2r/twxtyrm12ks7x98wpyccvdk00000gn/T/d-115111-4598-elq1su/package.tgz" "--target=0.21.0" "--arch=x64"
npm ERR! node v0.10.35
npm ERR! npm  v2.3.0
npm ERR! code ELIFECYCLE

npm ERR! pty.js@0.2.5 install: `node-gyp rebuild`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the pty.js@0.2.5 install script 'node-gyp rebuild'.
npm ERR! This is most likely a problem with the pty.js package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node-gyp rebuild
npm ERR! You can get their info via:
npm ERR!     npm owner ls pty.js
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /private/var/folders/2r/twxtyrm12ks7x98wpyccvdk00000gn/T/apm-install-dir-115111-4598-p25mkv/npm-debug.log

**Comment by stelek** [comment]
The problem is with "pty", which is external package...

**Comment by mdixon47** [comment]
How do we repair or replace the external package?

**Comment by stelek** [comment]
It may be worth filing a bug on "pty" - https://github.com/chjj/pty.js/. And then updating the dependancies.

**Comment by D-side** [comment]
I suspect it's already over there: chjj/pty.js#103

**Comment by stelek** [comment]
Looks like it...

**Comment by mdixon47** [comment]
More infomation that could be useful. 

npm ERR! Please include the following file with any support request:
npm ERR!     /private/var/folders/2r/twxtyrm12ks7x98wpyccvdk00000gn/T/apm-install-dir-115116-14661-12b0v59/npm-debug.log

**Comment by D-side** [comment]
@mdixon47 in future, read carefully: *include the following __file__*. Not a path to it. Cheers :)

The issue is already identified (I think?).

**Comment by mdixon47** [comment]
@D-side Please forgive my stupidity. For future references, I will include the file if I know the location of the problem. I just would like to have the problem fixed and not reading what I should or should not have included. 

@D-side (Paul) if you are unable to fix the problem. Direct me to someone who can.

**Comment by Stevo5o** [comment]
@mdixon47 hi, have to you had any luck on the install or is there anyone there with some ideas on how to fix this? Thanks

apm install term2 with Windows PowerShell.
C:\Windows\system32> apm -v
apm  0.137.0
npm  2.3.0
node 0.10.35
python 2.7.9
git 1.9.5.msysgit.0
visual studio 2013

C:\Windows\system32> apm install term2
Installing term2 to C:\Users\Stephen\.atom\packages failed
> pty.js@0.2.5 install C:\Users\Stephen\AppData\Local\Temp\apm-install-dir-115118-8064-1icwy4m\node_modules\term2\node_modules\pt
y.js
> node-gyp rebuild
C:\Users\Stephen\AppData\Local\Temp\apm-install-dir-115118-8064-1icwy4m\node_modules\term2\node_modules\pty.js>node "C:\Users\Ste
phen\AppData\Local\atom\app-0.179.0\resources\app\apm\node_modules\npm\bin\node-gyp-bin\\..\..\node_modules\node-gyp\bin\node-gyp
.js" rebuild
gyp: deps\winpty\winpty.gyp not found (cwd: C:\Users\Stephen\AppData\Local\Temp\apm-install-dir-115118-8064-1icwy4m\node_modules\
term2\node_modules\pty.js) while loading dependencies of binding.gyp while trying to load binding.gyp
gypnpm ERR! Windows_NT 6.2.9200
npm ERR! argv "C:\\Users\\Stephen\\AppData\\Local\\atom\\app-0.179.0\\resources\\app\\apm\\bin\\\\node.exe" "C:\\Users\\Stephen\\
AppData\\Local\\atom\\app-0.179.0\\resources\\app\\apm\\node_modules\\npm\\bin\\npm-cli.js" "--globalconfig" "C:\\Users\\Stephen\
\.atom\\.apm\\.apmrc" "--userconfig" "C:\\Users\\Stephen\\.atom\\.apmrc" "install" "C:\\Users\\Stephen\\AppData\\Local\\Temp\\d-1
15118-8064-1nf8165\\package.tgz" "--target=0.21.0" "--arch=ia32" "--msvs_version=2013"
npm ERR! node v0.10.35
npm ERR! npm  v2.3.0
npm ERR! code ELIFECYCLE

npm ERR! pty.js@0.2.5 install: `node-gyp rebuild`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the pty.js@0.2.5 install script 'node-gyp rebuild'.
npm ERR! This is most likely a problem with the pty.js package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node-gyp rebuild
npm ERR! You can get their info via:
npm ERR!     npm owner ls pty.js
npm ERR! There is likely additional logging output above.

**Comment by D-side** [comment]
@Stevo5o @mdixon47 I've referenced an issue on `pty.js` above, it suggests that `pty.js` (that `term2` is based on) is not yet compatible with `io.js` that recent versions of Atom use.

At least, release notes for v0.177.0 claim that *"Atom now runs on top of Chrome 40 and io.js"*, it's kind of odd that `node 0.10.35` is reported by APM (in Atom 0.179.0 too). An explaination would be appreciated.

So I guess until `pty.js` supports `io.js`, nothing can be done on user side but waiting.

**I really wish to be wrong now.**

**Comment by Stevo5o** [comment]
OK thanks @D-side. 
I am going back to using Netbeans for certain jobs. For me I will keep atom for writing Javascript and having fun with GitHub. Maybe if I was running Atom on Linux, it would be easier to configure.  Netbeans has some nice features and for windows it's quite easy to configure for chrome, Powershell, node, bootstrap, Java, PHP Git etc.

**Comment by mdixon47** [comment]
@D-side I will be patient. Atom is an excellent tool for coding hands down. I will work on it until a solution is found.

**Comment by clov3rly** [comment]
@chjj may have just fixed the issue upstream chjj/pty.js@3ceabf9b but this is still refusing to build here. Anyone know if apm/npm is caching the dependencies, package.json points at master, so it should work now?

**Comment by brutal-chaos** [comment]
nan is now failing. pty.js depends on 1.6.2, however it looks like 1.7.0 is needed. Using pty.js with nan 1.7.0, atom-term2 built, but failed to run.

**Comment by brutal-chaos** [comment]
correction: I had done something incorrectly. Simply having pty.js depend on nan 1.7.0 works

**Comment by flogball00** [comment]
changed package.json in pty.js to nan 1.7.0 doesnt quite do it for me.

```
 pty.js@0.2.5 install /private/var/folders/rm/rc7vn00s3q35229jrjm9blpm0000gn/T/apm-install-dir-11525-57490-o57q65/node_modules/term2/node_modules/pty.js
> node-gyp rebuild

  CXX(target) Release/obj.target/pty/src/unix/pty.o

In file included from ../src/unix/pty.cc:16:
../node_modules/nan/nan.h:67:21: error: no type named 'ExternalAsciiStringResource' in 'v8::String'; did you mean 'ExternalStringResource'?
typedef v8::String::ExternalAsciiStringResource
```

**Comment by brutal-chaos** [comment]
@flogball00 are you pointing atom-term2 to the modified pty.js? I got it to build fine.

![term2](https://cloud.githubusercontent.com/assets/443413/6520436/697e6458-c37a-11e4-99bc-1054c908ca1e.png)

**Comment by flogball00** [comment]
still trying to install it.  did "apm install term2" and get the errors.  Am i missing something?

**Comment by thejayvm** [comment]
What `package.json` file are you referring to? I see that each package has it's own, but I have no `term2` directory in `~/.atom/packages` to change anything in.

**Comment by brutal-chaos** [comment]
@thejayvm fork pty.js and change it's nan dependency to 1.7.0. Then have atom-term2's package.json pty.js dependency point to your fork. `apm install` and all will work. Or just wait for @chjj to bump the nan dependency.

**Comment by thejayvm** [comment]
@brutal-chaos Thanks! Do I fork term2 as well? I'm still unclear as to how I get the `atom-term2` package directory.

**Comment by flogball00** [comment]
clone the atom-term2 repo into /.atom/packages/ and change the package.json

**Comment by thejayvm** [comment]
@flogball00 @brutal-chaos Thanks for the help folks. I've got it working now.

However my colours are no longer working... anyone else have issues with this?

**Comment by brutal-chaos** [comment]
I have color, but am unable to change them in the settings. This should probably be moved to a new issue.

**Comment by stelek** [comment]
Rebuilding with new reps work like a charm. Hope it will be fixed on master

**Comment by brutal-chaos** [comment]
@f this (and similar issues) can now be closed. Waiting on pty.js to change their nan-dependency to 1.7.0 is all that's needed to get term2 working again.

**Comment by D-side** [comment]
New `pty.js` (0.2.6) has been released 5 days ago, which uses the new `nan`.

This issue can now be closed. `term2` installs just fine.

**Comment by mdixon47** [comment]
Thank you every for your time and patience in the matter.  Thank you D-side for your information. It was very helpful. As D-Side says, This issue is now closed.
